### DISCUSSION TITLE
ESPNow on the ESP32-S3

### DISCUSSION THREAD

**Discussion by brokenbrigde9000** [micropython/micropython#13613] [ESP32]
I have previously used ESPNow on an ESP32 which worked pretty well. I wanted to send some data between ESP32-S3 modules but ESPNow does not seem to be present in the latest builds. I'm using ESP32-S3-WROOM-1U-N16R8 units with firmware v1.22.1. I can import espnow, but performing e=espnow.ESPNow() does not give me anything. dir(espnow) shows me an emtpy package too. The preview build v1.23.0 gives me the same result. Is this funtionality not present in the S3 firmware?

-edit
I just reuploaded v1.22.1 and presto: espnow.ESPnow exists. Then I rebooted the ESP32-S3 and what do you know, its gone! Nowhere to be found. I rebooted again a couple of times and saw a espnow.wifi once when doing dir(espnow), but nothing else. Why did espnow disappear while it did manage to show up once?

**Comment by ned-pcs** [comment]
Are you using the version that is set up for the octal SPIRAM? If not, give it a try.

**Reply by brokenbrigde9000** [reply]
Yes, i'm using the SPIRAM version. micropthon.mem_info() shows me that the memory has initialized properly.

**Comment by peterhinch** [comment]
I did some testing a few days ago using the UM Feather S3 boards with V1.22.1. ESPnow worked fine. These boards have QSPI SPIRAM. If your board has octal SPIRAM could this be the problem?

**Reply by brokenbrigde9000** [reply]
I don't know, can it? I used the 'support for Octal-SPIRAM version of Micropython so I assumed that would work as such.

**Reply by ned-pcs** [reply]
The N16R8 modules have octal SPIRAM. I'm using these. I have used the stock Octal-SPIRAM builds in the past.

You want to make sure that you use the `.bin` version, rather than the `.app.bin` version, and do an `erase_flash` first:
```sh
esptool.py --chip esp32s3 --port /dev/ttyACM1 erase_flash
esptool.py --chip esp32s3 --port /dev/ttyACM1 write_flash -z 0 ESP32_GENERIC_S3-SPIRAM_OCT-20240207-v1.23.0-preview.91.g5a68e82d1.bin
```
That version worked for me, and has the `espnow` code:
```
>>> import espnow
>>> dir(espnow)
['__class__', '__name__', 'ADDR_LEN', 'ESPNowBase', 'KEY_LEN', 'MAX_DATA_LEN', 'MAX_ENCRYPT_PEER_NUM', 'MAX_TOTAL_PEER_NUM', '__dict__', '__file__', 'ESPNow']
```

**Reply by brokenbrigde9000** [reply]
Thank you, it is working now. I don't quite know why, I thought I uploaded the .bin before but it might have been the .uf2 instead. I used Thonny to upload the image (manual selection of image, no automatic choice by Thonny) which has given me some problems before.

I used the .bin as you said and used esptool directly to upload it. Now  esp.ESPNow continues to exist after a reboot. Great!

**Reply by brokenbrigde9000** [reply]
Ok, its not working again and the reason is as simple as it is dumb: I copied a folder to the ESP called 'espnow'. If you call import espnow, it doesn't import the espnow I meant by that but starts looking for stuff in that folder which isn't there.

Solution: Don't use a folder called espnow.
